### ISSUE TITLE
Component keeps connection port used after it goes to exception

### ISSUE THREAD

**Issue by joaobrittoneto** [Brazilian-Institute-of-Robotics/drivers-orogen-usbl_evologics#28]
An error is occurring this last days with the driver.
Once the driver goes to exception, it tries to reconnect with the usbl.
But its not able to open the socket, saying it is already in use.
After that we get Gb of `[ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9`

Why the FileDescriptorActivity keeps trying to execute operation in the file descriptor once the component goes to exception?
Once the component goes to exception, the iodrivers_base/rtt should not release the socket?

@doudou, @g-arjones, any idea?

```
    1 0.063 [ Warning][TaskContextServer()] CTaskContext 'usbl_auv' already bound to CORBA Naming Service.
    2 0.063 [ Warning][TaskContextServer()] Trying to rebind... done. New CTaskContext bound to Naming Service.
    3 0.064 [ Warning][TaskContextServer()] CTaskContext 'usbl_auv_Logger' already bound to CORBA Naming Service.
    4 0.065 [ Warning][TaskContextServer()] Trying to rebind... done. New CTaskContext bound to Naming Service.
    5 [20171121-10:04:54:862] [ WARN] - iodrivers_base::FD given to Driver::setFileDescriptor is set as blocking, setting the NONBLOCK flag (/home/flatfish/dev.bir/drivers/iodrivers_base/src/IOStream.cpp:25 - iodrivers_base::FDStream::FDS      tream(int, bool))
    6 20.560 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
    7 21.060 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
    8 21.561 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
    9 22.061 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
   10 22.065 [ ERROR  ][CDataFlowInterface_i::createConnection] in configure(): switching to exception state because of unhandled exception
   11 22.065 [ ERROR  ][CDataFlowInterface_i::createConnection]   readPacket(): no data after retrying with remaining time 20000ms of 20000ms timeout
   12 22.562 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
   13 23.062 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
   14 23.154 [ ERROR  ][CDataFlowInterface_i::createConnection] in configure(): switching to exception state because of unhandled exception
   15 23.154 [ ERROR  ][CDataFlowInterface_i::createConnection]   cannot open server socket on port 49155: Address already in use
   16 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: timeout in select()
   17 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   18 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   19 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   20 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   21 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   22 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   23 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   24 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   25 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9
   26 23.563 [ ERROR  ][CDataFlowInterface_i::createConnection] FileDescriptorActivity: error in select(), errno = 9

**Comment by g-arjones** [comment]
>  Once the component goes to exception, the iodrivers_base/rtt should not release the socket?

This is not rtt's responsability since it doesn't own the socket.

What is intriguing for me is why/how the activity is still trying to manipulate the fd once the component is in exception state..

**Comment by doudou** [comment]
Activities are active from the start. So the fact that it runs is not unexpected.

The problem IMO is pretty much analoguous to the NULL discussion. configureHook() is absolutely not exception-safe. The scary bit being that it's not caused to something in this particular component, but due to accepted guidelines related to `iodrivers_base::Task`.

One MUST call setDriver / the superclass's configureHook early in a configureHook, in order to get the ability to talk to the device through the orocos ports. However, this means that the FileDescriptorActivity is already setup to listen to the FD. If there's an exception in the device configuration itself, the FDA stays initialized and the device doesn't get closed.

Ideally, I would go to remove the setup code out of iodrivers_base::Task::configureHook, to separate the different bits of setup (port setup, activity setup and setDriver), but I don't see how that could be done in a backward-compatible way.

An alternative would be to create a ConfigureGuard class in the iodrivers_base orogen component that could be used to migrate configureHooks to a saner flow, ensuring that the component gets properly de-initialized if the configure fails.

**Comment by joaobrittoneto** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Investigate USBL is blowing out of space our hard disk](https://trello.com/c/Wv4WlonW/164-investigate-usbl-is-blowing-out-of-space-our-hard-disk)

**Comment by g-arjones** [comment]
Now, I am confused.

1. My understanding of orocos tasks was that once the component enters an error state, the execution engine was stopped. Am I wrong about that? What is the point of keeping the execution engine running while the component is very likely in an inconsistent state?

2. Unless I got your point wrong, we shouldn't be experiencing what is happening in the second exception in the logs. From what I can see there, the exception happened during `openURI()`, that is, before `setDriver()`, before `iodrivers_base::Task::configureHook()` is called and therefore before the FileDescriptorActivity is setup.

3. Though I like the idea of the scope guard in `iodrivers_base::Task::configureHook()`, how would this help in this particular scenario where this method doesn't even get called. Also, wouldn´t it be appropriate to have rollback/cleanup stuff in errorHook() instead?

I thought I had a fair understanding of what was going on in RTT's engine but it seems that are quite a few bits I am still missing to get to that point..

**Comment by doudou** [comment]
> Am I wrong about that? What is the point of keeping the execution engine running while the component is very likely in an inconsistent state?

Operations can be called regardless of the state of the component, which means that an activity must be running.

> Unless I got your point wrong, we shouldn't be experiencing what is happening in the second exception in the logs.

My analysis was:
- `readPacket(): no data after retrying with remaining time 20000ms of ...` happens in configureHook after setDriver(), leaving the component in a bad state
- `cannot open server socket on port 49155: Address already in use` happens in `openURI`, but somehow breaks the "old" FD. I didn't think about it further, so I might have taken a shortcut I wasn't allowed to take here.

> Though I like the idea of the scope guard in iodrivers_base::Task::configureHook(), how would this help in this particular scenario where this method doesn't even get called

Well ... I obviously didn't consider the "this method doesn't even get called" scenario. Components are deconfigured before they enter exceptionHook (go through stopHook+cleanupHook if in RUNNING and cleanupHook if in STOPPED). When configureHook throws, the component is in PRE_OPERATIONAL state which means that no other handlers are called. However, none should be, since the contract is that configureHook must ensure that it leaves the task the way it found it.

In summary: we need the guard *and* we need to find out if that's truly the problem here

**Comment by g-arjones** [comment]
> Operations can be called regardless of the state of the component, which means that an activity must be running.

This makes sense.. But, I suppose no user code (updateHook()) will be called unless the component is RUNNING, right? So, why keep polling/watching the fds? This is obviously not the underlying problem since configureHook should be exception-safe in the first place but just something to consider.

> My analysis was

Now that I went through the logs again, I believe you are right. I don't know where did the idea that `TaskBase::configureHook()` didn't get called come from but that clearly wasn't the case since the component DID try to communicate with the device and that only happens after the base task's configureHook() is called.

> but somehow breaks the "old" FD.

The iodriver_base's dtor closes the fd and since this component is using a smart pointer, I believe it is safe to say that this was how the fd got invalidated.

It is all clear now. Other than unwatching the underlying fd do you see anything else that should be rolledback if configureHook throws?

**Comment by doudou** [comment]
> So, why keep polling/watching the fds? This is obviously not the underlying problem since configureHook should be exception-safe in the first place but just something to consider.

Activities are really "trigger mechanisms", that are used in other places than TaskContexts. The Activity is completely separated from the TaskContext, i.e. the former has no knowledge of the latter's state (and it should be kept this way). It's the overall system integration's responsibility to ensure that no FDs are watched when the task is stopped.
